File: javascript/217.contains-duplicate.js
/*
 * @lc app=leetcode id=217 lang=javascript
 *
 * [217] Contains Duplicate
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {boolean}
 */
var containsDuplicate = function (nums) {
  // Solution 1:
  return new Set(nums).size !== nums.length;

  // Solution 2:
  // for (let i = 0; i < nums.length; i++) {
  //   if (nums.indexOf(nums[i], i + 1) !== -1) return true;
  // }
  // return false;

  // Solution 3:
  // const appear = {};
  // for (let i = 0; i < nums.length; i++) {
  //   if (appear[nums[i]]) return true;
  //   appear[nums[i]] = true;
  // }
  // return false;
};
// @lc code=end

File: rust/217.contains-duplicate.rs
/*
 * @lc app=leetcode id=217 lang=rust
 *
 * [217] Contains Duplicate
 */

// @lc code=start
impl Solution {
    pub fn contains_duplicate(nums: Vec<i32>) -> bool {
        // Solution 1: sort then compare
        // let mut sorted_nums = nums.to_vec();
        // sorted_nums.sort_unstable();
        // for i in 0..(sorted_nums.len() - 1) {
        //     if sorted_nums[i] == sorted_nums[i + 1] {
        //         return true;
        //     }
        // }

        // Solution 2: HashSet
        use std::collections::HashSet;
        let mut num_set: HashSet<i32> = HashSet::with_capacity(nums.len());
        for i in 0..nums.len() {
            if num_set.contains(&nums[i]) {
                return true;
            }
            num_set.insert(nums[i]);
        }

        return false;
    }
}
// @lc code=end
